docs/en/development/architecture.md

```diff
@@ -182,6 +182,43 @@ No matter what pool is used for a job, at start `ThreadStatus` instance is creat
 
+## Concurrency control {#concurrency-control}
+Query that can be parallelized uses `max_threads` setting to limit itself. Default value for this setting is selected in a way that allows single query to utilize all CPU cores in the best way. But what if there are multiple concurrent queries and each of them uses default `max_threads` setting value? Then queries will share CPU resources. OS will ensure fairness by constantly switching threads, which introduce some performance penalty. `ConcurrencyControl` helps to deal with this penalty and avoid allocating a lot of threads. Configuration setting `concurrent_threads_soft_limit_num` is used to limit how many concurrent thread can be allocated before applying some kind of CPU pressure.
+
+:::note
+`concurrent_threads_soft_limit_num` and `concurrent_threads_soft_limit_ratio_to_cores` are disabled (equal 0) by default. So this feature must be enabled before use.
+:::
+
+Notion of cpu `slot` is introduced. Slot is unit of concurrency: to run a thread query has to acquire a slot in advance and release it when thread stops. Amount of slots is globally limited in a server. Multiple concurrent queries are competing for cpu slots if total demand exceed total amount of slots. `ConcurrencyControl` is responsible to resolve this competition by doing cpu slot scheduling in a fair manner.
+
+Each slot can be seen as independent state machine with the following states:
+ * `free`: slot is available to be allocated by any query.
+ * `granted`: slot is `allocated` by specific query, but not yet acquired by any thread.
+ * `acquired`: slot is `allocated` by specific query and acquired by a thread.
+
+Note that `allocated` slot can be in two different states: `granted` and `acquired`. The former is transitional state, that actually should be short. From instant when slot is allocated to a query till the moment when up-scaling procedure is run by any thread of that query.
+
+```mermaid
+stateDiagram-v2
+    direction LR
+    [*] --> free
+    free --> allocated: allocate
+    state allocated {
+        direction LR
+        [*] --> granted
+        granted --> acquired: acquire
+        acquired --> [*]
+    }
+    allocated --> free: release
+```
+
+API of `ConcurrencyControl` consists of the following functions:
+1. Create an allocation for a query: `auto slots = ConcurrencyControl::instance().allocate(1, max_threads);`. It will allocate at least 1 and at most `max_threads` slots. Note that the first slot is granted immediately, but remaining slots may be granted later. Thus limit is soft, because every query will obtain at least one thread.
```

Comment:
```suggestion
1. Create a resource allocation for a query: `auto slots = ConcurrencyControl::instance().allocate(1, max_threads);`. It will allocate at least 1 and at most `max_threads` slots. Note that the first slot is granted immediately, but the remaining slots may be granted later. Thus limit is soft, because every query will obtain at least one thread.
```
Maybe "resource allocation" is better here?